Repo Index --merge-repo Panics

k kudo repo index ~/repo -w --merge-repo community
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1e29426]

goroutine 1 [running]:
github.com/kudobuilder/kudo/pkg/kudoctl/util/repo.NewClient(0x0, 0x0, 0x0, 0x0)
	/Users/alenavarkockova/go/src/github.com/kudobuilder/kudo/pkg/kudoctl/util/repo/repo_operator.go:42 +0x26
github.com/kudobuilder/kudo/pkg/kudoctl/cmd.(*repoIndexCmd).run(0xc000326080, 0xc000101600, 0x1)
	/Users/alenavarkockova/go/src/github.com/kudobuilder/kudo/pkg/kudoctl/cmd/repo_index.go:135 +0x152
github.com/kudobuilder/kudo/pkg/kudoctl/cmd.newRepoIndexCmd.func1(0xc0001da780, 0xc000101600, 0x1, 0x4, 0x0, 0x0)
	/Users/alenavarkockova/go/src/github.com/kudobuilder/kudo/pkg/kudoctl/cmd/repo_index.go:70 +0x94
github.com/spf13/cobra.(*Command).execute(0xc0001da780, 0xc0001015c0, 0x4, 0x4, 0xc0001da780, 0xc0001015c0)
	/Users/alenavarkockova/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000db400, 0x0, 0x2240fa0, 0xc00009c058)
	/Users/alenavarkockova/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/alenavarkockova/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
	/Users/alenavarkockova/go/src/github.com/kudobuilder/kudo/cmd/kubectl-kudo/main.go:24 +0x27


** --merge <url> works **

--- a/pkg/kudoctl/cmd/install/install.go
+++ b/pkg/kudoctl/cmd/install/install.go
@@ -100,7 +100,7 @@ func installOperator(operatorArgument string, options *Options, fs afero.Fs, set
 	if err != nil {
 		return errors.WithMessage(err, "could not build operator repository")
 	}
-	clog.V(4).Printf("repository used %v", repository)
+	clog.V(4).Printf("repository used %s", repository)
 
 	kc, err := kudo.NewClient(settings.Namespace, settings.KubeConfig)
 	clog.V(3).Printf("acquiring kudo client")

--- a/pkg/kudoctl/cmd/repo_index.go
+++ b/pkg/kudoctl/cmd/repo_index.go
@@ -112,7 +112,6 @@ func (ri *repoIndexCmd) run() error {
 		if config == nil {
 			return fmt.Errorf("configuration for repositories does not contain %s", ri.urlRepoName)
 		}
-		fmt.Printf("url: %v", config.URL)
 		ri.url = config.URL
 	}
 
@@ -167,7 +166,6 @@ func merge(index *repo.IndexFile, mergeIndex *repo.IndexFile) {
 
 func (ri *repoIndexCmd) mergeRepoConfig() (*repo.Configuration, error) {
 	if ri.mergeRepoName != "" {
-		fmt.Printf("repo name %q\n", ri.mergeRepoName)
 		return ri.repoConfig(ri.mergeRepoName)
 	}
 

--- a/pkg/kudoctl/cmd/repo_index.go
+++ b/pkg/kudoctl/cmd/repo_index.go
@@ -105,7 +105,7 @@ func (ri *repoIndexCmd) validate(args []string) error {
 func (ri *repoIndexCmd) run() error {
 
 	if ri.urlRepoName != "" {
-		config, err := ri.repoConfig()
+		config, err := ri.repoConfig(ri.urlRepoName)
 		if err != nil {
 			return err
 		}
@@ -167,7 +167,8 @@ func merge(index *repo.IndexFile, mergeIndex *repo.IndexFile) {
 
 func (ri *repoIndexCmd) mergeRepoConfig() (*repo.Configuration, error) {
 	if ri.mergeRepoName != "" {
-		return ri.repoConfig()
+		fmt.Printf("repo name %q\n", ri.mergeRepoName)
+		return ri.repoConfig(ri.mergeRepoName)
 	}
 
 	return &repo.Configuration{
@@ -176,10 +177,10 @@ func (ri *repoIndexCmd) mergeRepoConfig() (*repo.Configuration, error) {
 	}, nil
 }
 
-func (ri *repoIndexCmd) repoConfig() (*repo.Configuration, error) {
+func (ri *repoIndexCmd) repoConfig(repoName string) (*repo.Configuration, error) {
 	repos, err := repo.LoadRepositories(ri.fs, Settings.Home.RepositoryFile())
 	if err != nil {
 		return nil, err
 	}
-	return repos.GetConfiguration(ri.urlRepoName), nil
+	return repos.GetConfiguration(repoName), nil
 }

--- a/pkg/kudoctl/util/repo/repo.go
+++ b/pkg/kudoctl/util/repo/repo.go
@@ -51,7 +51,8 @@ func NewRepositories() *Repositories {
 }
 
 // GetConfiguration returns a RepoName Config for a name or nil
-func (r Repositories) GetConfiguration(name string) *Configuration {
+func (r *Repositories) GetConfiguration(name string) *Configuration {
+	fmt.Printf("%v\n", r.Repositories)
 	for _, repo := range r.Repositories {
 		if repo.Name == name {
 			return repo
@@ -61,7 +62,7 @@ func (r Repositories) GetConfiguration(name string) *Configuration {
 }
 
 // CurrentConfiguration provides the repo config for the current context
-func (r Repositories) CurrentConfiguration() *Configuration {
+func (r *Repositories) CurrentConfiguration() *Configuration {
 	return r.GetConfiguration(r.Context)
 }
 

--- a/pkg/kudoctl/util/repo/repo.go
+++ b/pkg/kudoctl/util/repo/repo.go
@@ -3,6 +3,7 @@ package repo
 import (
 	"fmt"
 	"os"
+	"strings"
 
 	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1alpha1"
 	"github.com/kudobuilder/kudo/pkg/kudoctl/kudohome"
@@ -28,11 +29,30 @@ type Configuration struct {
 	Name string `json:"name"`
 }
 
+// Configurations is a collection of Configuration for Stringer
+type Configurations []*Configuration
+
 // Repositories represents the repositories.yaml file usually in the $KUDO_HOME
 type Repositories struct {
-	RepoVersion  string           `json:"repoVersion"`
-	Context      string           `json:"context"`
-	Repositories []*Configuration `json:"repositories"`
+	RepoVersion  string         `json:"repoVersion"`
+	Context      string         `json:"context"`
+	Repositories Configurations `json:"repositories"`
+}
+
+// String is a stringer function for Configuration
+func (c *Configuration) String() string {
+	return fmt.Sprintf("{ name:%v, url:%v }", c.Name, c.URL)
+}
+
+// String is a stringer function for Configurations
+func (c Configurations) String() string {
+
+	confs := make([]string, len(c))
+	for i, config := range c {
+		confs[i] = config.String()
+	}
+
+	return fmt.Sprintf("repo configs: %v", strings.Join(confs, ","))
 }
 
 // Default initialized repository.

--- a/pkg/kudoctl/util/repo/repo_operator.go
+++ b/pkg/kudoctl/util/repo/repo_operator.go
@@ -27,6 +27,10 @@ type Client struct {
 	Client http.Client
 }
 
+func (c *Client) String() string {
+	return c.Config.String()
+}
+
 // ClientFromSettings retrieves the operator repo for the configured repo in settings
 func ClientFromSettings(fs afero.Fs, home kudohome.Home, repoName string) (*Client, error) {
 	rc, err := ConfigurationFromSettings(fs, home, repoName)
@@ -53,9 +57,9 @@ func NewClient(conf *Configuration) (*Client, error) {
 }
 
 // DownloadIndexFile fetches the index file from a repository.
-func (r *Client) DownloadIndexFile() (*IndexFile, error) {
+func (c *Client) DownloadIndexFile() (*IndexFile, error) {
 	var indexURL string
-	parsedURL, err := url.Parse(r.Config.URL)
+	parsedURL, err := url.Parse(c.Config.URL)
 	if err != nil {
 		return nil, errors.Wrap(err, "parsing config url")
 	}
@@ -63,7 +67,7 @@ func (r *Client) DownloadIndexFile() (*IndexFile, error) {
 
 	indexURL = parsedURL.String()
 
-	resp, err := r.Client.Get(indexURL)
+	resp, err := c.Client.Get(indexURL)
 	if err != nil {
 		return nil, errors.Wrap(err, "getting index url")
 	}
@@ -81,10 +85,10 @@ func (r *Client) DownloadIndexFile() (*IndexFile, error) {
 // The PackageVersion is a package configuration from the index file which has a list of urls where
 // the package can be pulled from.  This will cycle through the list of urls and will return the reader
 // from the first successful url.  If all urls fail, the last error will be returned.
-func (r *Client) getPackageReaderByAPackageURL(pkg *PackageVersion) (*bytes.Buffer, error) {
+func (c *Client) getPackageReaderByAPackageURL(pkg *PackageVersion) (*bytes.Buffer, error) {
 	var pkgErr error
 	for _, u := range pkg.URLs {
-		r, err := r.getPackageBytesByURL(u)
+		r, err := c.getPackageBytesByURL(u)
 		if err == nil {
 			return r, nil
 		}
@@ -95,9 +99,9 @@ func (r *Client) getPackageReaderByAPackageURL(pkg *PackageVersion) (*bytes.Buff
 	return nil, pkgErr
 }
 
-func (r *Client) getPackageBytesByURL(packageURL string) (*bytes.Buffer, error) {
+func (c *Client) getPackageBytesByURL(packageURL string) (*bytes.Buffer, error) {
 	clog.V(4).Printf("attempt to retrieve package from url: %v", packageURL)
-	resp, err := r.Client.Get(packageURL)
+	resp, err := c.Client.Get(packageURL)
 	if err != nil {
 		return nil, errors.Wrap(err, "getting package url")
 	}
@@ -106,11 +110,11 @@ func (r *Client) getPackageBytesByURL(packageURL string) (*bytes.Buffer, error) 
 }
 
 // GetPackageBytes provides an io.Reader for a provided package name and optional version
-func (r *Client) GetPackageBytes(name string, version string) (*bytes.Buffer, error) {
+func (c *Client) GetPackageBytes(name string, version string) (*bytes.Buffer, error) {
 	clog.V(4).Printf("getting package reader for %v, %v", name, version)
-	clog.V(5).Printf("repository using: %v", r.Config)
+	clog.V(5).Printf("repository using: %v", c.Config)
 	// Construct the package name and download the index file from the remote repo
-	indexFile, err := r.DownloadIndexFile()
+	indexFile, err := c.DownloadIndexFile()
 	if err != nil {
 		return nil, errors.WithMessage(err, "could not download repository index file")
 	}
@@ -120,12 +124,12 @@ func (r *Client) GetPackageBytes(name string, version string) (*bytes.Buffer, er
 		return nil, errors.Wrapf(err, "getting %s in index file", name)
 	}
 
-	return r.getPackageReaderByAPackageURL(pkgVersion)
+	return c.getPackageReaderByAPackageURL(pkgVersion)
 }
 
 // GetPackage provides an Package for a provided package name and optional version
-func (r *Client) GetPackage(name string, version string) (packages.Package, error) {
-	reader, err := r.GetPackageBytes(name, version)
+func (c *Client) GetPackage(name string, version string) (packages.Package, error) {
+	reader, err := c.GetPackageBytes(name, version)
 	if err != nil {
 		return nil, err
 	}

